File: api/src/main/java/com/ruby/rubia_server/core/controller/AILogController.java
package com.ruby.rubia_server.core.controller;

import com.ruby.rubia_server.core.dto.CreateAILogDTO;
import com.ruby.rubia_server.core.dto.UpdateAILogDTO;
import com.ruby.rubia_server.core.entity.AILog;
import com.ruby.rubia_server.core.enums.AILogStatus;
import com.ruby.rubia_server.core.service.AILogService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.format.annotation.DateTimeFormat;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

@RestController
@RequestMapping("/api/ai-logs")
@RequiredArgsConstructor
@Slf4j
public class AILogController {

    private final AILogService aiLogService;

    @PostMapping
    public ResponseEntity<AILog> createAILog(@Valid @RequestBody CreateAILogDTO createDTO) {
        log.info("Creating new AI log for company: {}", createDTO.getCompanyId());
        AILog aiLog = aiLogService.createAILog(createDTO);
        return ResponseEntity.status(HttpStatus.CREATED).body(aiLog);
    }

    @GetMapping("/{id}")
    public ResponseEntity<AILog> getAILogById(@PathVariable UUID id) {
        log.info("Fetching AI log with ID: {}", id);
        Optional<AILog> aiLog = aiLogService.getAILogById(id);
        return aiLog.map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @GetMapping
    public ResponseEntity<Page<AILog>> getAllAILogs(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "10") int size,
            @RequestParam(defaultValue = "createdAt") String sortBy,
            @RequestParam(defaultValue = "desc") String sortDir) {
        
        Sort sort = sortDir.equalsIgnoreCase("desc") ? 
                Sort.by(sortBy).descending() : 
                Sort.by(sortBy).ascending();
        
        Pageable pageable = PageRequest.of(page, size, sort);
        Page<AILog> aiLogs = aiLogService.getAllAILogs(pageable);
        return ResponseEntity.ok(aiLogs);
    }

    @GetMapping("/company/{companyId}")
    public ResponseEntity<List<AILog>> getAILogsByCompanyId(@PathVariable UUID companyId) {
        log.info("Fetching AI logs for company: {}", companyId);
        List<AILog> aiLogs = aiLogService.getAILogsByCompanyId(companyId);
        return ResponseEntity.ok(aiLogs);
    }

    @GetMapping("/status/{status}")
    public ResponseEntity<List<AILog>> getAILogsByStatus(@PathVariable AILogStatus status) {
        log.info("Fetching AI logs with status: {}", status);
        List<AILog> aiLogs = aiLogService.getAILogsByStatus(status);
        return ResponseEntity.ok(aiLogs);
    }

    @GetMapping("/ai-agent/{aiAgentId}")
    public ResponseEntity<List<AILog>> getAILogsByAIAgentId(@PathVariable UUID aiAgentId) {
        log.info("Fetching AI logs for AI agent: {}", aiAgentId);
        List<AILog> aiLogs = aiLogService.getAILogsByAIAgentId(aiAgentId);
        return ResponseEntity.ok(aiLogs);
    }

    @PutMapping("/{id}")
    public ResponseEntity<AILog> updateAILog(@PathVariable UUID id, @Valid @RequestBody UpdateAILogDTO updateDTO) {
        log.info("Updating AI log with ID: {}", id);
        Optional<AILog> updatedAILog = aiLogService.updateAILog(id, updateDTO);
        return updatedAILog.map(ResponseEntity::ok)
                .orElse(ResponseEntity.notFound().build());
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteAILog(@PathVariable UUID id) {
        log.info("Deleting AI log with ID: {}", id);
        boolean deleted = aiLogService.deleteAILog(id);
        return deleted ? ResponseEntity.noContent().build() : ResponseEntity.notFound().build();
    }

    @GetMapping("/company/{companyId}/total-cost")
    public ResponseEntity<BigDecimal> getTotalCostByCompanyId(@PathVariable UUID companyId) {
        log.info("Calculating total cost for company: {}", companyId);
        BigDecimal totalCost = aiLogService.getTotalCostByCompanyId(companyId);
        return ResponseEntity.ok(totalCost);
    }

    @GetMapping("/company/{companyId}/total-tokens")
    public ResponseEntity<Long> getTotalTokensUsedByCompanyId(@PathVariable UUID companyId) {
        log.info("Calculating total tokens used for company: {}", companyId);
        Long totalTokens = aiLogService.getTotalTokensUsedByCompanyId(companyId);
        return ResponseEntity.ok(totalTokens);
    }

    @GetMapping("/company/{companyId}/count")
    public ResponseEntity<Long> countAILogsByCompanyIdAndStatus(
            @PathVariable UUID companyId,
            @RequestParam AILogStatus status) {
        log.info("Counting AI logs for company: {} with status: {}", companyId, status);
        long count = aiLogService.countAILogsByCompanyIdAndStatus(companyId, status);
        return ResponseEntity.ok(count);
    }

    @GetMapping("/date-range")
    public ResponseEntity<List<AILog>> getAILogsByDateRange(
            @RequestParam @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime startDate,
            @RequestParam @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME) LocalDateTime endDate) {
        log.info("Fetching AI logs between {} and {}", startDate, endDate);
        List<AILog> aiLogs = aiLogService.getAILogsByDateRange(startDate, endDate);
        return ResponseEntity.ok(aiLogs);
    }
}

File: api-dotnet/Rubia.Server/Controllers/AILogController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Rubia.Server.DTOs;
using Rubia.Server.Entities;
using Rubia.Server.Enums;
using Rubia.Server.Services.Interfaces;

namespace Rubia.Server.Controllers;

[ApiController]
[Route("api/ai-logs")]
[Authorize]
public class AILogController : ControllerBase
{
    private readonly IAILogService _aiLogService;
    private readonly ILogger<AILogController> _logger;

    public AILogController(IAILogService aiLogService, ILogger<AILogController> logger)
    {
        _aiLogService = aiLogService;
        _logger = logger;
    }

    [HttpPost]
    public async Task<ActionResult<AILogDto>> CreateAILog([FromBody] CreateAILogDto createDto)
    {
        try
        {
            if (!ModelState.IsValid)
            {
                return BadRequest(ModelState);
            }

            _logger.LogInformation("Creating new AI log for company: {CompanyId}", createDto.CompanyId);
            var aiLog = await _aiLogService.CreateAILogAsync(createDto);
            var dto = MapToDto(aiLog);
            
            return CreatedAtAction(nameof(GetAILogById), new { id = aiLog.Id }, dto);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error creating AI log");
            return StatusCode(500, "Internal server error");
        }
    }

    [HttpGet("{id:guid}")]
    public async Task<ActionResult<AILogDto>> GetAILogById(Guid id)
    {
        try
        {
            _logger.LogInformation("Fetching AI log with ID: {Id}", id);
            var aiLog = await _aiLogService.GetAILogByIdAsync(id);
            
            if (aiLog == null)
            {
                return NotFound();
            }

            return Ok(MapToDto(aiLog));
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching AI log {Id}", id);
            return StatusCode(500, "Internal server error");
        }
    }

    [HttpGet]
    public async Task<ActionResult<IEnumerable<AILogDto>>> GetAllAILogs(
        [FromQuery] int page = 0,
        [FromQuery] int size = 10,
        [FromQuery] string sortBy = "createdAt",
        [FromQuery] string sortDir = "desc")
    {
        try
        {
            var aiLogs = await _aiLogService.GetAllAILogsAsync(page, size, sortBy, sortDir);
            var dtos = aiLogs.Select(MapToDto);
            
            return Ok(dtos);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching AI logs");
            return StatusCode(500, "Internal server error");
        }
    }

    [HttpGet("company/{companyId:guid}")]
    public async Task<ActionResult<IEnumerable<AILogDto>>> GetAILogsByCompanyId(Guid companyId)
    {
        try
        {
            _logger.LogInformation("Fetching AI logs for company: {CompanyId}", companyId);
            var aiLogs = await _aiLogService.GetAILogsByCompanyIdAsync(companyId);
            var dtos = aiLogs.Select(MapToDto);
            
            return Ok(dtos);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching AI logs for company {CompanyId}", companyId);
            return StatusCode(500, "Internal server error");
        }
    }

    [HttpGet("status/{status}")]
    public async Task<ActionResult<IEnumerable<AILogDto>>> GetAILogsByStatus(AILogStatus status)
    {
        try
        {
            _logger.LogInformation("Fetching AI logs with status: {Status}", status);
            var aiLogs = await _aiLogService.GetAILogsByStatusAsync(status);
            var dtos = aiLogs.Select(MapToDto);
            
            return Ok(dtos);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching AI logs by status {Status}", status);
            return StatusCode(500, "Internal server error");
        }
    }

    [HttpGet("ai-agent/{aiAgentId:guid}")]
    public async Task<ActionResult<IEnumerable<AILogDto>>> GetAILogsByAIAgentId(Guid aiAgentId)
    {
        try
        {
            _logger.LogInformation("Fetching AI logs for AI agent: {AiAgentId}", aiAgentId);
            var aiLogs = await _aiLogService.GetAILogsByAIAgentIdAsync(aiAgentId);
            var dtos = aiLogs.Select(MapToDto);
            
            return Ok(dtos);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching AI logs for AI agent {AiAgentId}", aiAgentId);
            return StatusCode(500, "Internal server error");
        }
    }

    [HttpPut("{id:guid}")]
    public async Task<ActionResult<AILogDto>> UpdateAILog(Guid id, [FromBody] UpdateAILogDto updateDto)
    {
        try
        {
            if (!ModelState.IsValid)
            {
                return BadRequest(ModelState);
            }

            _logger.LogInformation("Updating AI log with ID: {Id}", id);
            var updatedAILog = await _aiLogService.UpdateAILogAsync(id, updateDto);
            
            if (updatedAILog == null)
            {
                return NotFound();
            }

            return Ok(MapToDto(updatedAILog));
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error updating AI log {Id}", id);
            return StatusCode(500, "Internal server error");
        }
    }

    [HttpDelete("{id:guid}")]
    public async Task<IActionResult> DeleteAILog(Guid id)
    {
        try
        {
            _logger.LogInformation("Deleting AI log with ID: {Id}", id);
            var deleted = await _aiLogService.DeleteAILogAsync(id);
            
            if (!deleted)
            {
                return NotFound();
            }

            return NoContent();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error deleting AI log {Id}", id);
            return StatusCode(500, "Internal server error");
        }
    }

    [HttpGet("company/{companyId:guid}/total-cost")]
    public async Task<ActionResult<decimal>> GetTotalCostByCompanyId(Guid companyId)
    {
        try
        {
            _logger.LogInformation("Calculating total cost for company: {CompanyId}", companyId);
            var totalCost = await _aiLogService.GetTotalCostByCompanyIdAsync(companyId);
            
            return Ok(totalCost);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error calculating total cost for company {CompanyId}", companyId);
            return StatusCode(500, "Internal server error");
        }
    }

    [HttpGet("company/{companyId:guid}/total-tokens")]
    public async Task<ActionResult<long>> GetTotalTokensUsedByCompanyId(Guid companyId)
    {
        try
        {
            _logger.LogInformation("Calculating total tokens used for company: {CompanyId}", companyId);
            var totalTokens = await _aiLogService.GetTotalTokensUsedByCompanyIdAsync(companyId);
            
            return Ok(totalTokens);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error calculating total tokens for company {CompanyId}", companyId);
            return StatusCode(500, "Internal server error");
        }
    }

    [HttpGet("company/{companyId:guid}/count")]
    public async Task<ActionResult<long>> CountAILogsByCompanyIdAndStatus(
        Guid companyId,
        [FromQuery] AILogStatus status)
    {
        try
        {
            _logger.LogInformation("Counting AI logs for company: {CompanyId} with status: {Status}", companyId, status);
            var count = await _aiLogService.CountAILogsByCompanyIdAndStatusAsync(companyId, status);
            
            return Ok(count);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error counting AI logs for company {CompanyId} with status {Status}", companyId, status);
            return StatusCode(500, "Internal server error");
        }
    }

    [HttpGet("date-range")]
    public async Task<ActionResult<IEnumerable<AILogDto>>> GetAILogsByDateRange(
        [FromQuery] DateTime startDate,
        [FromQuery] DateTime endDate)
    {
        try
        {
            _logger.LogInformation("Fetching AI logs between {StartDate} and {EndDate}", startDate, endDate);
            var aiLogs = await _aiLogService.GetAILogsByDateRangeAsync(startDate, endDate);
            var dtos = aiLogs.Select(MapToDto);
            
            return Ok(dtos);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error fetching AI logs by date range {StartDate} - {EndDate}", startDate, endDate);
            return StatusCode(500, "Internal server error");
        }
    }

    private static AILogDto MapToDto(AILog aiLog)
    {
        return new AILogDto
        {
            Id = aiLog.Id,
            CompanyId = aiLog.CompanyId,
            AiAgentId = aiLog.AiAgentId,
            UserId = aiLog.UserId,
            ConversationId = aiLog.ConversationId,
            MessageId = aiLog.MessageId,
            MessageTemplateId = aiLog.MessageTemplateId,
            RequestPrompt = aiLog.RequestPrompt,
            RawResponse = aiLog.RawResponse,
            ProcessedResponse = aiLog.ProcessedResponse,
            TokensUsedInput = aiLog.TokensUsedInput,
            TokensUsedOutput = aiLog.TokensUsedOutput,
            EstimatedCost = aiLog.EstimatedCost,
            Status = aiLog.Status,
            ErrorMessage = aiLog.ErrorMessage,
            CreatedAt = aiLog.CreatedAt
        };
    }
}